### DISCUSSION TITLE
Manual bass issue

### DISCUSSION THREAD

**Discussion by newVideophile** [GrandOrgue/grandorgue#1425] [Help]
I will play at a wedding in a month with two keyboards and GO. For some pratical reasons, I won't have a pedalboard. 

Therefore I will use a very usefull function called "Manual Bass" on Johannus organs. On GO, it can be obtained using the coupling panel, then the great (Hauptwerk), and selecting "BAS" in the Pedal section: 
![Screenshot_20230305_213627_VNC Viewer](https://user-images.githubusercontent.com/95248680/222984737-b0b64f8c-50c6-4fd7-8e54-9fabe0be57d5.jpg)

In theory, the lower note played on the great organ should be played on the pedalboard as well. 

However, when I play, there are many missed notes at the pedalboard, and thus "holes" in my piece. 

Here is for example a situation when it will fail to play the lowest note all the time: 
First a C major chord: C E G, then I shift my hand to the right by one white key, to get a D minor chord (D F A). In this case the C will be played on the pedalboard, but not the D.

If I play a C alone and then a D, it works, even if I play "more than legato", i.e. with the C and D played simultaneously for a split second.

If I play the two same chords, but in the opposite order (DFA, then CEG), it works !!! 

The only way I have found so far to get what I want is to detach the bass, but from a musical standpoint, this is not what I want!

So, either I missing something, or I have found a bug...

**Comment by oleg68** [comment]
@newVideophile could you test a workaround: assign the same midi events to the first manual and to the pedal and stop using the coupling panel in GO?

**Comment by newVideophile** [comment]
This doesn't work, because all notes played at the first manual would be played at the pedal and this is not what I want. The "BAS" coupling is supposed to play only the lower note of the manual on the pedal.

**Comment by larspalo** [comment]
Both the BAS and MEL coupler will sometimes "fail", and _depending on the manner of playing_ it will always be possible to create problems and issues that won't be desired from a musical standpoint (notes jumping around that are not intended to be part of the bass line or melody line for instance). Please realise that the BAS and MEL couplers generally are _**very**_ poor choices to emulate multiple manuals for anything more demanding than a simple hymn that has more or less equal note values in all voices. I'm able to fool (and thus make them fail musically) all kinds of commercial digital instruments that have such features built in - one way or another...

That said - if you have any concrete ideas of how to improve the algorithms used for BAS and MEL couplers - feel free to submit them.

**Comment by kerkovits** [comment]
@newVideophile When using the BAS coupler, I usually accommodate my playing style, and always release bass notes just a tiny moment before moving to the next chord. It is less disturbing that the bass line is not perfect legato.

@larspalo In the church, we have a Viscount organ with an automatic pedal function. Its algorithm is different from the one of GO. GO checks on keypress if it is the lowest note, and does not override this decision later (misses the bass note if pressed too early). Viscount checks _all_ notes upon _every_ keypress: if the bass note of the new chord was missed due to early keypress, but the alto or tenor keypress comes later (after releasing the bass of the previous chord), it saves the situation. Pro: Less bass notes will be missing. Con: This algorithm assumes that there are no rests in the bass line.

I usually find the results of the Viscount algorithm better suited to my (and possibly most organists') playing style, but it is also not perfect (the other organist in our church can frequently loose bass notes even on the Viscount instrument).

**Comment by newVideophile** [comment]
Yes, releasing the bass is also the only workaround I have found, although it is very unnatural for me to play this way, as I tend to detach too much (in the fear of missing notes) and the result is not what I would like.

I the church where I play, they have a Johannus and I don't know its algorithm, but it just works. I will try to experiment the next time I will be playing on it (in 3 weeks from now)

**Reply by kerkovits** [reply]
If you avoid using characteristic ranks (e.g. Posaune 32') and use only smooth pipes (e.g. Subbaß 16'), the effect of the missing bass will be so little that no one except organists would notice it. In one of the church buildings, there is only a digital piano, and we use that with GO, BAS coupler enabled. No one has ever asked me about missing notes in the bass. 😉 So do not be afraid, this 'non-legato' bass can be only subtly different from true legato.

**Comment by larspalo** [comment]
So, I've tested modifying the BAS/MEL couple targeting a little. The result from it is that I feel this is very much like the proverb that a choice must be made between plague or cholera...

Compared to the current implementation the changes I tried does improve the hit rate quite a lot and sounds pretty good when the bass and the melody is played with a continuous perfect legato leaving no gaps at all. When playing simultaneous detached chords it works fine if the releases are exact, but very quickly the BAS and MEL starts jumping around to the closest note if the bass or top note isn't held or played perfectly legato and that sounds - well, not so good either, and in some ways much worse than the present implementation.

If there's a very high interest to test this I can push the test branch to my fork and it should be possible to get a public build to test, but to be honest, I'm not really convinced that it's any real improvement over the current state - it just behaves slightly differently, which might suit some people better and some other worse...

**Reply by kerkovits** [reply]
I would be happy to try it if it is not a significant work for you.

**Reply by larspalo** [reply]
It's available at https://github.com/larspalo/grandorgue/actions/runs/4363784520 if anyone wants to test it. The way I changed it there will always be a BAS or MEL note coupled at all times - which means: do not hold a chord in any middle voice if the BAS or MEL voice has any rest or show the slightest of gaps... Also this means that over-legato (note commune) will work for both BAS and MEL, but at the price mentioned above!

**Reply by newVideophile** [reply]
Well, I think that I prefer missed notes to unwanted notes. 

My understanding is that if I play a chord BDF, then CEG, but with a slight overlapping, i.e. BDF, then for a split second BCEG, and finally CEG, the current alorithm will replicate the B on the pedal but not the C, because C is not seen as a new bass.

So, I have a suggestion for an improved algorithm, based on the current one, but with a tweak. 

Here is the tweak: when the bass is released, check if the new "candidate bass", which is the new lowest note played, has been pressed less than 50 ms ago. If yes, accept it as the new bass. If not, play no bass. Of course, the 50 ms should probably be adpated, it could even be a user setting. 

With my example:
t = 0 ms, BDF -> bass = B
t = 1000 ms, BCEG -> bass = B
t = 1020 ms, CEG -> new bass candidate C, previous bass released less than 50 ms ago? -> yes (20 ms) -> bass = C

**Comment by larspalo** [comment]
Warning! This is a long post! Read only if you're interested in understanding why the missed notes happen and want to understand the current BAS/MEL coupler algorithm.

Since we probably all can agree that my "un-failing" experiment version actually fails worse than the current implementation, but I've nevertheless done some research in the code and there seems to be a lack of understanding of how the BAS/MEL coupler algorithm work I'll walk through the relevant part of the code here as mainly academic documentation.

The relevant part of the code can presently be found in src/grandorgue/model/GOCoupler.cpp at line 288 but before I present the code with my (rather lengthy) comments explaining what happens some background information might be necessary.

This function is actually called every time a manual sets a key (actually via GOCoupler::SetKey). This is happening whether the coupler is active or not, the actual decision to act on the key presses (or releases) happen in GOCoupler::SetOut.

For each coupler object there are three vectors that has the size of the number of logical keys to handle, each of them built up exactly the same (but used for different reasons). The first logical key is the element with index zero. Here we are mainly interested in the vector named m_KeyVelocity which contain the input values from the manual for this coupler object. The unsigned values themselves are key velocities, where 0 is equal to a note off.

```
// input values are every keypress/release with note value starting at 0 for
// the first key, remember that a release (note off) also is sent with
// velocity value of 0 and will be processed
void GOCoupler::ChangeKey(int note, unsigned velocity) {
  if (m_UnisonOff)
    return; // if the coupler is an unison off we quit immediately
  if (m_CouplerType == COUPLER_BASS || m_CouplerType == COUPLER_MELODY) {
    // this part is the relevant one for BAS and MEL couplers
    // the first thing that happens is a scan of the m_KeyVelocity vector
    // that contain information of which keys are pressed, meaning they have
    // a value that's not zero. This scan always finds the lowest or highest
    // active key - it's never failing
    int nextNote = -1;
    if (m_CouplerType == COUPLER_BASS) {
      // for a BAS coupler the scan is done from the bottom - first/lowest key
      // and upwards until either a pressed key is found or the end of the
      // vector is reached
      for (nextNote = 0; nextNote < (int)m_KeyVelocity.size(); nextNote++)
        if (m_KeyVelocity[nextNote] > 0)
          break;
      if (nextNote == (int)m_KeyVelocity.size())
        nextNote = -1;
    } else {
      // for a MEL coupler the scan is done from the top - highest key and
      // downwards until a pressed key is found or the first key is checked
      for (nextNote = m_KeyVelocity.size() - 1; nextNote >= 0; nextNote--)
        if (m_KeyVelocity[nextNote] > 0)
          break;
    }

    // the m_CurrentTone is the used note to couple and what the next if
    // statement does is to cancel the current note if it's not the found
    // lowest or highest active note anymore the value of -1 means that
    // no key is used - keys are numbered from 0, remember!
    if (m_CurrentTone != -1 && nextNote != m_CurrentTone) {
      SetOut(m_CurrentTone + m_Keyshift, 0);
      m_CurrentTone = -1;
    }

    // The first part of next if statement checks if the velocity is above zero
    // which means that the note sent to this function is an attack and not a
    // release and if that note also is the lowest/highest detected in the scan
    // and thus should be the active note.
    // The other part after the or sign becomes true when the note sent into
    // this function is a release and the note detected in the scan was the
    // last attack. I'm highly doubtful if this ever can happen so that the
    // result in m_CurrentTone is anything else than -1 which again means that
    // the coupler is not affecting any key
    if (((velocity > 0 && nextNote == note)
         || (velocity == 0 && nextNote == m_LastTone)))
      m_CurrentTone = nextNote;

    // If the m_CurrentTone is actually affecting any key then that information
    // is sent to SetOut which will populate the outgoing vector that contain
    // information sent to the manual that this coupler is targeting
    if (m_CurrentTone != -1)
      SetOut(m_CurrentTone + m_Keyshift, m_KeyVelocity[m_CurrentTone]);

    // if the note sent to this function is an attack it will be remembered
    // as m_LastTone - but this is only the last attack and any single release
    // will wipe it away by setting the value to -1
    if (velocity > 0)
      m_LastTone = note;
    else
      m_LastTone = -1;
    return;
  }
  // Down here is the only thing that happens for a normal coupler. The input
  // values are simply forwarded to SetOut where it will end up in the output
  // vector if values are valid and coupler is active.
  SetOut(note + m_Keyshift, velocity);
}
```

The reason why a note is missed in some circumstances is because the previous BAS or MEL note is not released and will "conceal" the new intended note, either in upwards movement (for BAS) or in downwards movements (for MEL) and at the same time other notes are pressed (and possibly released) which will erase the m_LastTone information.

Here is a simple debugging output of a note missing scenario for a BAS coupler. The 048-C and 067-G are struck together with BAS coupler correctly triggering and then 052-E and 064-E are played at which point the BAS coupling is lost. All of this happen very quickly.

```
(The attack of the high g is detected first)
Input was note: 31 with velocity: 339
m_CurrentTone: 31
nextNote: 31
m_LastTone: 31

(next the attack of the low c which correctly trigger the BAS coupler)
Input was note: 12 with velocity: 323
m_CurrentTone: 12
nextNote: 12
m_LastTone: 12

(then the lower e is detected - but note that the current active BAS note is still the lower c)
Input was note: 16 with velocity: 331
m_CurrentTone: 12
nextNote: 12
m_LastTone: 16

(and then the high e is detected - note that the high g is not release yet either == I played with slight overlegato!)
Input was note: 28 with velocity: 375
m_CurrentTone: 12
nextNote: 12
m_LastTone: 28

(Next some releases are detected. First out is the high g, at this point the low c is still the active note)
Input was note: 31 with velocity: 0
m_CurrentTone: 12
nextNote: 12
m_LastTone: -1

(Here the low c is released which will set m_CurrentTone to -1 even though we can see that the scan correctly "see" that e should be the next note but the last note is also -1 and the triggering note on that happened three events earlier is long forgotten and the BAS coupling is missed)
Input was note: 12 with velocity: 0
m_CurrentTone: -1
nextNote: 16
m_LastTone: -1

(Then the lower e is released which only leaves the higher e to be detected but it won't be triggered either as it's three events away)
Input was note: 16 with velocity: 0
m_CurrentTone: -1
nextNote: 28
m_LastTone: -1

(Finally the high e is released)
Input was note: 28 with velocity: 0
m_CurrentTone: -1
nextNote: -1
m_LastTone: -1
```

So, as long as upper notes are held and only lower notes change, one can play legato in the bass any direction. If any upper note changes too - the only way to be fairly sure to trigger the BAS in upwards movement is by playing non-legato (detached) in the bass.

I've tried many different versions of implementation, but none I've tested so far is performing better in every aspect than the current one.

**Reply by newVideophile** [reply]
Very interesting, thank you for sharing!

Did you add the comments or were they already included in the code (if this is the case, I'm impressed)?

So, I have two questions:
 - I dont' understand your comment about (velocity == 0 && nextNote == m_LastTone): "I'm highly doubtful if this ever can happen". I don't see it as improbable. For example, if you play only one note at a time on the manual but "overlegato", this condition will be true and it will ensure that a bass is always coupled.
 - What about my suggestion of yesterday for an improved algorithm? Following the current code logic, it would require a m_KeyTime[] table containing the time at which each key was last pressed, and replace the (velocity == 0 && nextNote == m_LastTone) condition with:
   (velocity == 0 && (nextNote==-1 ? false : ((now-m_KeyTime[nextNote])<OVERLEGATO_MAX))
where "now" is the current time (actually the time when the current note was released) and OVERLEGATO_MAX corresponds to the 50 ms value I mentioned in my previous post.

**Reply by larspalo** [reply]
@newVideophile I've added all the comments as you can see if you compare to the code existing in the repository.

If you read carefully, I didn't say that it couldn't happen, but most of the time when it could happen m_LastTone will already be -1. And in the case of a single voice on the manual, all this is anyway not really needed.

No timing information is readily available as of now. A kind of queue could be used to determine in which order notes are played, but the overhead of handling it is most likely not really defendable.

**Reply by larspalo** [reply]
@newVideophile After a quick test I should change the comment indeed. The condition (velocity == 0 && nextNote == m_LastTone) is what enables legato playing upwards for a BAS coupler - provided that no other note will wipe away the m_LastTone information.

**Reply by newVideophile** [reply]
> No timing information is readily available as of now. A kind of queue could be used to determine in which order notes are played, but the overhead of handling it is most likely not really defendable.

I understand. I think that in a short term, I will have to adapt my playing to what GO can currently do. Then I will create my own manual bass algorithm on Arduino, since all my keyboards are Arduino controlled. In a microntroller environment, getting timing information is easy.

**Reply by larspalo** [reply]
@newVideophile I've spent many hours now trying to improve the algorithm in different ways, so far I've unfortunately not found the golden solution.

The current implementation suffer from not "remembering" the last bass candidate enough, which leads to missing notes when one plays too much legato at the "wrong" time.

It's fairly easy to improve the hit rate by keeping track of the candidates better (longer), but then comes the real tricky part of deciding when to invalidate the candidate (now it's done as soon as a key release is processed - which is much too indiscriminate). With every different implementation I've tested it seems that if I gain in one aspect (for instance less missed notes) - I loose in another (more not desired notes will creep into the playing under different conditions). I can tell you that it's quite frustrating!

**Comment by oleg68** [comment]
@larspalo Why not to search the new lowest pressed note every time when any note is pressed or is released?

**Reply by larspalo** [reply]
@oleg68 Yes, I've tried that (for BAS couplers) and it sure works to increase hit rate (and point to the correct next note to target instead of only the very latest) but it's so easy to get too much jumps of the bass part up to the next voice that's not intended. The real problem is finding the right logic that avoid most of the "wrong" jumps, but still allow for legato playing.

**Reply by larspalo** [reply]
@newVideophile @kerkovits @oleg68 I've pushed my latest work to the alterBasMelCoupler branch in my fork. I almost think that it's an improvement over current BAS/MEL implementation, but when the build is finished please test it and if you have ideas for further improvements (or better implementation overall) feel free to present them.

**Reply by oleg68** [reply]
> but it's so easy to get too much jumps of the bass part up to the next voice that's not intended.

Why is it a problem? The intermediate bass notes are pressed for a very short time when usually bass pipes have a long attack, so they are released before their attack have finished (of cause, it works whell only with `Release sample scaling`).

If it is not desired, we may introduce the tracker delay for the bass coupler and play the bass note only if had not been released until the delay time passed.

**Reply by larspalo** [reply]
@oleg68 It's a big musical problem that's very apparent if you really have tested it with any serious playing.

As I showed with my first commit, it's very easy to make the BAS/MEL coupler never miss a note - but it's behaving musically very badly that way as it will jump to the next available note as soon as the bass or melody has the slightest gap - in effect not at all respecting voice leading or when the bass/melody actually should have a rest.

Again, the problem with the current implementation is that it will "forget" the next targeted note if there is more than one release or keypress in between them. Thus the bass or melody will sometimes miss notes that are intended to be part of that voice. This is also a musical problem, albeit actually less than never missing a note.

What I've tried to do is to keep a better track of what is the next possible candidate of the BAS/MEL voice and if the right conditions occur it will select that note. Missed notes are more or less a history. The selection/rejection criteria makes for better voice leading, but on some occasions there might still be unwanted jumps because it's impossible to differ between if the player intended to go to the next note that's pressed or the player simply just lifted the previous key. I solved that so that if the last key pressed in between current coupled note not is the next candidate, it will be invalidated if the bass/melody is lifted.

Any form of simple timing selection will ultimately fail to detect what the real intention of the player was - only actually analysing the keypress behaviour and taking into account all that has happened on the keyboard, a reasonably accurate prediction can be made.

**Reply by kerkovits** [reply]
I plan to test it tomorrow. Thank you.

**Reply by kerkovits** [reply]
PS. I agree with @larspalo that it would be a very bad idea to have short jumps in the BAS coupler, so the simple implementation is worse than the current one.

**Reply by kerkovits** [reply]
I tested it.

Compared to the present implementation, the hit rate was 100% with my playing style. (I tested only simple 4-part chorales.) Congratulations.

The biggest drawback becomes apparent, when there are repeating notes in the bass or the soprano. I explain the latter one:
1. I press the chord. In some cases, accidentally the alto is pressed a few ms later than the soprano -> alto becomes a candidate note for the MEL coupler. I press this chord for a long time, but alto is still a candidate note
2. The soprano note is repeated in the next chord, so I must release it after ca. 3/4 of the written time. All other notes stay pressed to have a "legato effect".
3. For the remaining time the MEL coupler jumps to the alto note.
Similar issue if the bass note is repeated, and the tenor was pressed a few ms later than the bass.

I am now wondering, on other ideas: using the original implementation but instead of:
```
    if (((velocity > 0 && nextNote == note)
         || (velocity == 0 && nextNote == m_LastTone)))
```
writing simply:
```
    if (((velocity > 0)
         || (velocity == 0 && nextNote == m_LastTone)))
```
Would increase the hit rate but not to 100%, while those 'jumps' described above would never occur.

Just thinking aloud, maybe `m_LastTone` should not set to -1, if `velocity==0`, but keep its original value.

These are just ideas.

**Reply by larspalo** [reply]
@kerkovits If you repeat the notes of MEL or BAS voice then the jumps do not occur. Nor does if happen if the bas/mel voice was the last to move.

But try to find an algorithm that can differ between the overlegato of a bass note ascending and the movement of a tenor voice! To the computer the ascending bass and the moving tenor is the same thing as long as the previous bass note still sounds. Think about it and you'll see.

Feel free to test your idea with the code. I've tested more or less everything I could imagine and I can tell you that your suggestion will not work to eliminate missed notes. Why? I leave that up to you to find out.

After working on this for about three days, I'm actually very doubtful if there ever can be an algorithm that will understand exactly what I as a player intended to be the bass or melody in every possible case that can occur. In real life I've never played on an instrument with this type of feature that I couldn't fool into doing the "wrong" thing musically one way or another...

I won't create any PR from my last commit to the official GO repository if not enough people think it's worth it, as I myself am not convinced and anyway think the BAS/MEL couplers are of questionable value compared to real pedal or true separate manual playing. For me it was more of an academic challenge. I'd gladly like to see someone else providing a better solution for BAS/MEL, but I've spent enough time with it for now!

**Reply by oleg68** [reply]
> I won't create any PR from my last commit to the official GO repository if not enough people think it's worth it, as I myself am not convinced and anyway think the BAS/MEL couplers are of questionable value compared to real pedal or true separate manual playing.

I find you comments in the GOCoupler.cpp very useful for understanding. I'd glad to see your PR just with your comments.

**Comment by newVideophile** [comment]
It seems that the need for "BAS" and "MEL" is not new: [https://www.youtube.com/watch?v=0EtW5HqlNIg](https://www.youtube.com/watch?v=0EtW5HqlNIg)

The algorithm is very simple here, but no software is involved ;-)
